features/cryptocell/FEATURE_CRYPTOCELL310/ccm_alt.c

```diff
@@ -109,31 +133,76 @@ int mbedtls_ccm_auth_decrypt( mbedtls_ccm_context *ctx, size_t length,
                       const unsigned char *tag, size_t tag_len )
 
 {
-    CRYSError_t CrysRet = CRYS_OK;
+    CRYSError_t crys_ret = CRYS_OK;
+    int ret = 0;
     /*
      * Check length requirements: SP800-38C A.1
      * Additional requirement: a < 2^16 - 2^8 to simplify the code.
      * 'length' checked later (when writing it to the first block)
      */
     if( tag_len < 4 || tag_len > 16 || tag_len % 2 != 0 )
-        return ( MBEDTLS_ERR_CCM_BAD_INPUT );
+        return( MBEDTLS_ERR_CCM_BAD_INPUT );
 
     /* Also implies q is within bounds */
     if( iv_len < 7 || iv_len > 13 )
-        return ( MBEDTLS_ERR_CCM_BAD_INPUT );
+        return( MBEDTLS_ERR_CCM_BAD_INPUT );
 
 #if SIZE_MAX > UINT_MAX
     if( length > 0xFFFFFFFF || add_len > 0xFFFFFFFF )
-        return ( MBEDTLS_ERR_CCM_BAD_INPUT );
+        return( MBEDTLS_ERR_CCM_BAD_INPUT );
 #endif
 
-    CrysRet =  CRYS_AESCCM( SASI_AES_DECRYPT, ctx->cipher_key, ctx->keySize_ID,(uint8_t*)iv, iv_len,
+    crys_ret =  CRYS_AESCCM( SASI_AES_DECRYPT, ctx->cipher_key, ctx->key_size,(uint8_t*)iv, iv_len,
                             (uint8_t*)add, add_len,  (uint8_t*)input, length, output, tag_len, (uint8_t*)tag );
-    if ( CrysRet != CRYS_OK )
-        return ( MBEDTLS_ERR_CCM_HW_ACCEL_FAILED );
+    if( crys_ret == CRYS_AESCCM_ILLEGAL_PARAMETER_SIZE_ERROR )
+    {
+       /*
+        * When CRYS_AESCCM_ILLEGAL_PARAMETER_SIZE_ERROR is returned,
+        * no operation has occured, and no need to zeroize output.
+        * In addition, it could be that the message length is too big,
+        * returning this error code, and we don't want to overflow
+        * the output buffer.
+        */
+       return( MBEDTLS_ERR_CCM_BAD_INPUT );
+    }
+    else if( crys_ret == CRYS_FATAL_ERROR )
+    {
+        /*
+         * Unfortunately, Crys AESCCM returns CRYS_FATAL_ERROR when
+         * MAC isn't as expected.
+         */
+        ret = MBEDTLS_ERR_CCM_AUTH_FAILED;
+        goto exit;
+    }
+    else if( crys_ret != CRYS_OK )
+    {
+        ret = MBEDTLS_ERR_PLATFORM_HW_ACCEL_FAILED;
+        goto exit;
+    }
+
+exit:
+    if( ret != 0 )
+        mbedtls_platform_zeroize( output, length );
+    return( ret );
 
-    return ( 0 );
+}
 
+int mbedtls_ccm_star_encrypt_and_tag( mbedtls_ccm_context *ctx, size_t length,
+                         const unsigned char *iv, size_t iv_len,
+                         const unsigned char *add, size_t add_len,
+                         const unsigned char *input, unsigned char *output,
+                         unsigned char *tag, size_t tag_len )
+{
+    return( MBEDTLS_ERR_PLATFORM_FEATURE_UNSUPPORTED );
+}
+
+int mbedtls_ccm_star_auth_decrypt( mbedtls_ccm_context *ctx, size_t length,
+                      const unsigned char *iv, size_t iv_len,
+                      const unsigned char *add, size_t add_len,
+                      const unsigned char *input, unsigned char *output,
+                      const unsigned char *tag, size_t tag_len )
```

Comment:
The parameter alignment seems wrong here in these two functions

Reply:
ok, but this is same is in the [origin](https://github.com/ARMmbed/mbedtls/blob/development/library/ccm.c#L307)

Reply:
Probably okay to leave it then.

Reply:
no problem, I'll fix it